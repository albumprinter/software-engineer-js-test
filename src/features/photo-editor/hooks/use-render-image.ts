import { useCallback } from 'react';

import { CANVAS_WIDTH, CANVAS_HEIGHT } from '../constants';
import { ImageParams, Coordinates, ImageSize } from '../types';

type RenderImage = (params: ImageParams) => void;
type UseRenderImageParams = {
  canvasRef: React.RefObject<HTMLCanvasElement>;
  setImageCoordinates: React.Dispatch<React.SetStateAction<Coordinates>>;
  initialCoordinates: React.MutableRefObject<Coordinates>;
  setImageSize: React.Dispatch<React.SetStateAction<ImageSize>>;
};

/**
 * Returns the function for rendering an uploaded image.
 * @param {Object} params - Properties and functions for rendering an image.
 * @param {Object} params.canvasRef - Reference to canvas element.
 * @param {Object} params.initialCoordinates - Initial image coordinates.
 * @param {Function} params.setImageCoordinates - The x-axis and y-axis coordinates state setter.
 * @param {Function} params.setImageSize - An image width and height size state setter.
 * @returns {Function} - Function for rendering an image.
 */
export const useRenderImage = ({
  canvasRef,
  initialCoordinates,
  setImageCoordinates,
  setImageSize,
}: UseRenderImageParams): RenderImage =>
  useCallback<RenderImage>(
    ({
      image,
      xImageAxis,
      yImageAxis,
      isExported = false,
    }: ImageParams): void => {
      if (!canvasRef.current) return;

      const { current: canvas } = canvasRef;
      const ctx = canvas.getContext('2d');

      if (!ctx) return;

      const { width: imageWidth, height: imageHeight } = image;

      canvas.width = CANVAS_WIDTH;
      canvas.height = CANVAS_HEIGHT;

      /**
       * Drawing an image whose size exceeds canvas
       */
      if (imageWidth > CANVAS_WIDTH && imageHeight > CANVAS_HEIGHT) {
        // Centered coordinates in the destination canvas
        const imageCenterWidth = CANVAS_WIDTH / 2 - imageWidth / 2;
        const imageCenterHeight = CANVAS_HEIGHT / 2 - imageHeight / 2;

        /**
         * If an image is exported from the JSON file description than
         * exported coordinates is set otherwise an image is centered
         */
        const xAxis = isExported ? xImageAxis : imageCenterWidth;
        const yAxis = isExported ? yImageAxis : imageCenterHeight;

        // Draw an image onto the canvas
        ctx.drawImage(image, xAxis, yAxis, image.width, image.height);

        // Setting an image coordinates to the local state
        setImageCoordinates({
          xImageAxis: xAxis,
          yImageAxis: yAxis,
        });

        // Setting an image size to the local state
        setImageSize({
          imageWidth,
          imageHeight,
        });

        // Saving centered image coordinates for shift calculations
        initialCoordinates.current.xImageAxis = imageCenterWidth;
        initialCoordinates.current.yImageAxis = imageCenterHeight;

        return;
      }

      /**
       * Drawing an image whose width, size, or both fits into the canvas
       */
      // Max of the width and height ratios
      const scaleFactor = Math.max(
        CANVAS_WIDTH / imageWidth,
        CANVAS_HEIGHT / imageHeight
      );

      // Width and height based of the scale factor
      const scaledImageWidth = imageWidth * scaleFactor;
      const scaledImageHeight = imageHeight * scaleFactor;

      // A centered position coordinates for an image within the canvas
      const scaledXAxis = CANVAS_WIDTH / 2 - scaledImageWidth / 2;
      const scaledYAxis = CANVAS_HEIGHT / 2 - scaledImageHeight / 2;

      /**
       * If an image is exported from the JSON file description than exported
       * coordinates is set otherwise an image coordinates is based on scale factor
       */
      const xAxis = isExported ? xImageAxis : scaledXAxis;
      const yAxis = isExported ? yImageAxis : scaledYAxis;

      // Draw an image onto the canvas
      ctx.drawImage(image, xAxis, yAxis, scaledImageWidth, scaledImageHeight);

      // Setting an image coordinates to the local state
      setImageCoordinates({
        xImageAxis: xAxis,
        yImageAxis: yAxis,
      });

      // Setting an image size to the local state
      setImageSize({
        imageWidth: scaledImageWidth,
        imageHeight: scaledImageHeight,
      });

      // Saving centered image coordinates for shift calculations
      initialCoordinates.current.xImageAxis = scaledXAxis;
      initialCoordinates.current.yImageAxis = scaledYAxis;
    },
    [canvasRef, initialCoordinates, setImageCoordinates, setImageSize]
  );
