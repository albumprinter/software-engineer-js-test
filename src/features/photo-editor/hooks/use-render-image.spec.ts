import { renderHook, act } from "@testing-library/react-hooks";
import { useRenderImage } from "./use-render-image";
import "jest-canvas-mock";

const canvasElement = document.createElement("canvas");
const ctx = canvasElement?.getContext("2d");

const mockProps = {
  canvasRef: {
    current: canvasElement,
  },
  initialCoordinates: {
    current: { xImageAxis: 0, yImageAxis: 0 },
  },
  setImageCoordinates: jest.fn(),
  setImageSize: jest.fn(),
};

describe("useRenderImage", () => {
  it("should return handler for rendering an image", () => {
    const {
      result: { current: renderImage },
    } = renderHook(() => useRenderImage(mockProps));

    expect(typeof renderImage).toBe("function");
  });

  it("should not draw an image as canvas reference is null", () => {
    const renderImageProps = {
      image: new Image(320, 320),
      xImageAxis: 0,
      yImageAxis: 0,
      isExported: false,
    };

    const {
      result: { current: renderImage },
    } = renderHook(() =>
      useRenderImage({
        ...mockProps,
        canvasRef: {
          current: null,
        },
      })
    );

    act(() => {
      renderImage(renderImageProps);
    });

    const calls = ctx?.__getDrawCalls();

    expect(calls).toHaveLength(0);
  });

  it("should draw an image whose width, size, or both fits into the canvas", () => {
    const renderImageProps = {
      image: new Image(320, 320),
      xImageAxis: 0,
      yImageAxis: 0,
      isExported: false,
    };

    const {
      result: { current: renderImage },
    } = renderHook(() => useRenderImage(mockProps));

    act(() => {
      renderImage(renderImageProps);
    });

    const calls = ctx?.__getDrawCalls();

    expect(calls).toMatchSnapshot();
  });

  it("should draw an exported image which width, size or both fits into canvas", () => {
    const renderImageProps = {
      image: new Image(320, 320),
      xImageAxis: -48,
      yImageAxis: -48,
      isExported: true,
    };

    const {
      result: { current: renderImage },
    } = renderHook(() => useRenderImage(mockProps));

    act(() => {
      renderImage(renderImageProps);
    });

    const calls = ctx?.__getDrawCalls();

    expect(calls).toMatchSnapshot();
  });

  it("should draw an image whose size exceeds canvas", () => {
    const renderImageProps = {
      image: new Image(3024, 4032),
      xImageAxis: 0,
      yImageAxis: 0,
      isExported: false,
    };

    const {
      result: { current: renderImage },
    } = renderHook(() => useRenderImage(mockProps));

    act(() => {
      renderImage(renderImageProps);
    });

    const calls = ctx?.__getDrawCalls();

    expect(calls).toMatchSnapshot();
  });

  it("should draw an exported image which size exceeds canvas", () => {
    const renderImageProps = {
      image: new Image(3024, 4032),
      xImageAxis: -48,
      yImageAxis: -48,
      isExported: false,
    };

    const {
      result: { current: renderImage },
    } = renderHook(() => useRenderImage(mockProps));

    act(() => {
      renderImage(renderImageProps);
    });

    const calls = ctx?.__getDrawCalls();

    expect(calls).toMatchSnapshot();
  });
});
