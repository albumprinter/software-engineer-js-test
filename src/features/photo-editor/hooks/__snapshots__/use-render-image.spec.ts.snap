// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useRenderImage should draw an exported image which size exceeds canvas 1`] = `
[
  {
    "props": {
      "dHeight": 320,
      "dWidth": 320,
      "dx": 0,
      "dy": -240,
      "img": <img
        height="320"
        width="320"
      />,
      "sHeight": 320,
      "sWidth": 320,
      "sx": 0,
      "sy": 0,
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "drawImage",
  },
  {
    "props": {
      "dHeight": 320,
      "dWidth": 320,
      "dx": -48,
      "dy": -48,
      "img": <img
        height="320"
        width="320"
      />,
      "sHeight": 320,
      "sWidth": 320,
      "sx": 0,
      "sy": 0,
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "drawImage",
  },
  {
    "props": {
      "dHeight": 4032,
      "dWidth": 3024,
      "dx": -792,
      "dy": -1536,
      "img": <img
        height="4032"
        width="3024"
      />,
      "sHeight": 4032,
      "sWidth": 3024,
      "sx": 0,
      "sy": 0,
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "drawImage",
  },
  {
    "props": {
      "dHeight": 4032,
      "dWidth": 3024,
      "dx": -792,
      "dy": -1536,
      "img": <img
        height="4032"
        width="3024"
      />,
      "sHeight": 4032,
      "sWidth": 3024,
      "sx": 0,
      "sy": 0,
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "drawImage",
  },
]
`;

exports[`useRenderImage should draw an exported image which width, size or both fits into canvas 1`] = `
[
  {
    "props": {
      "dHeight": 320,
      "dWidth": 320,
      "dx": 0,
      "dy": -240,
      "img": <img
        height="320"
        width="320"
      />,
      "sHeight": 320,
      "sWidth": 320,
      "sx": 0,
      "sy": 0,
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "drawImage",
  },
  {
    "props": {
      "dHeight": 320,
      "dWidth": 320,
      "dx": -48,
      "dy": -48,
      "img": <img
        height="320"
        width="320"
      />,
      "sHeight": 320,
      "sWidth": 320,
      "sx": 0,
      "sy": 0,
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "drawImage",
  },
]
`;

exports[`useRenderImage should draw an image whose size exceeds canvas 1`] = `
[
  {
    "props": {
      "dHeight": 320,
      "dWidth": 320,
      "dx": 0,
      "dy": -240,
      "img": <img
        height="320"
        width="320"
      />,
      "sHeight": 320,
      "sWidth": 320,
      "sx": 0,
      "sy": 0,
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "drawImage",
  },
  {
    "props": {
      "dHeight": 320,
      "dWidth": 320,
      "dx": -48,
      "dy": -48,
      "img": <img
        height="320"
        width="320"
      />,
      "sHeight": 320,
      "sWidth": 320,
      "sx": 0,
      "sy": 0,
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "drawImage",
  },
  {
    "props": {
      "dHeight": 4032,
      "dWidth": 3024,
      "dx": -792,
      "dy": -1536,
      "img": <img
        height="4032"
        width="3024"
      />,
      "sHeight": 4032,
      "sWidth": 3024,
      "sx": 0,
      "sy": 0,
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "drawImage",
  },
]
`;

exports[`useRenderImage should draw an image whose width, size, or both fits into the canvas 1`] = `
[
  {
    "props": {
      "dHeight": 320,
      "dWidth": 320,
      "dx": 0,
      "dy": -240,
      "img": <img
        height="320"
        width="320"
      />,
      "sHeight": 320,
      "sWidth": 320,
      "sx": 0,
      "sy": 0,
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "drawImage",
  },
]
`;
